proposals/0046-astro-env.md

```diff
@@ -0,0 +1,359 @@
+- A public variable key has to be prefixed by `PUBLIC_`
+- A non public variable accessed using `import.meta.env` on the client side will be `undefined` (value will be accessible server side)
+- Env variables can be loaded through `.env` (or `.env.production`, `.env.development`) and CLI
+- Any non built-in variable can be [manually typed]([.env.production](https://docs.astro.build/en/guides/environment-variables/#intellisense-for-typescript))
+- Runtime variables should be access using `process.env`, or following the used runtime (eg. `Deno.env.get()` for the deno adapter)
+
+# Goals
+
+- Provide a fully type-safe experience for environment variables, without [manual type definitions](https://docs.astro.build/en/guides/environment-variables/#intellisense-for-typescript)
+- Reduce user confusion between inlined, static variables and dynamic, runtime variables
+- Allow adapters to specify how runtime env variables should be handled
+- Allow integrations to define environment variable constraints
+- Simple type casting and validation (strings/numbers/booleans, required/optional)
+
+# Non-Goals
+
+- Complex validation or type casting of env variables. We might want to enable this at some point, but there is likely a performance cost for this at runtime. We should punt on this if possible!
+- Future: allow adapters to customize build-time variable handling. Better runtime handling is the most important problem this proposal aims to solve, but if we find an API that enables this as well, that's great.
+
+# Detailed Design
+
+## Terminology
+
+- **Static variable**: variable replaced by its value at build time. Uses `import.meta.env` under the hood
+- **Dynamic variable**: variable retrieved at runtime, never part of the bundle. Uses runtime specific features, like `process.env` or `Deno.env.get()`
+
+## Astro config
+
+A new `env.schema` property is added to `AstroUserConfig`. Its type looks like so:
+
+```ts
+type StringField = {
+  type: "string"
+  optional?: boolean
+  default?: string
+}
+
+type NumberField = {
+  type: "number"
+  optional?: boolean
+  default?: number
+}
+
+type NumberField = {
+  type: "boolean"
+  optional?: boolean
+  default?: boolean
+}
+
+type EnvFieldType = StringField | NumberField | BooleanField
+
+type PublicStaticEnvFieldMetadata = {
+  scope: "static"
+  access: "public"
+}
+
+type PrivateStaticEnvFieldMetadata = {
+  scope: "static"
+  access: "private"
+}
+
+type PrivateDynamicEnvFieldMetadata = {
+  scope: "dynamic"
+  access: "private"
+}
+
+type EnvSchema =
+  | Record<`PUBLIC_${string}`, PublicStaticEnvFieldMetadata & EnvFieldType>
+  | Record<string, (PrivateStaticEnvFieldMetadata | PrivateDynamicEnvFieldMetadata) & EnvFieldType>
+
+type AstroUserConfig = {
+  env?: {
+    schema: EnvSchema
+  }
+}
+```
+
+Since there are various combinations possible that can make choosing confusing, we provide a `envField` helper that uses chaining, mirroring other APIs like Astro DB or zod:
```

Comment:
I wonder if it's cleaner to rename `envField` as `env` instead, but not a priority 😄

Reply:
Actually I've been thinking about this. The only issue I see with calling it `env` is that people could use it (because suggested by auto-complete) thinking it allows to access `env`. That's why I originally wanted to provide the helper from a function argument (see alternatives below) but ended up just renaming it to `envField`. I'm open to any naming tho! I think we can manipulate the ordering in the vscode extension too? Could be something to explore

Reply:
I think `envField` is sound. I see `env` too close to `process.**env**` 😅